src/main/java/net/minecraftforge/client/model/ModelDynBucket.java

```diff
@@ -0,0 +1,243 @@
+import net.minecraft.client.resources.model.ModelResourceLocation;
+import net.minecraft.item.ItemStack;
+import net.minecraft.util.EnumFacing;
+import net.minecraft.util.ResourceLocation;
+import net.minecraftforge.client.model.pipeline.UnpackedBakedQuad;
+import net.minecraftforge.fluids.Fluid;
+import net.minecraftforge.fluids.FluidContainerRegistry;
+import net.minecraftforge.fluids.FluidRegistry;
+import net.minecraftforge.fluids.FluidStack;
+
+import org.apache.commons.lang3.tuple.Pair;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+
+import javax.vecmath.Matrix4f;
+
+public class ModelDynBucket implements IModel {
+  public static final ModelResourceLocation LOCATION = new ModelResourceLocation(new ResourceLocation("forge", "dynbucket"), "inventory");
+
+  private static final float NORTH_Z_FLUID = 7.49f / 16f;
+  private static final float SOUTH_Z_FLUID = 8.51f / 16f;
+
+  private static final float NORTH_Z_COVER = 7.48f / 16f;
+  private static final float SOUTH_Z_COVER = 8.52f / 16f;
+
+  protected final IModel bucket;
+  protected static final ResourceLocation interior = new ResourceLocation("forge", "items/bucket_interior");
+  protected static final ResourceLocation front = new ResourceLocation("forge", "items/bucket_front");
+
+  private boolean hasTextures;
+
+  public ModelDynBucket(IModel bucket) {
+    this.bucket = bucket;
+  }
+
+  @Override
+  public Collection<ResourceLocation> getDependencies() {
+    return bucket.getDependencies();
+  }
+
+  @Override
+  public Collection<ResourceLocation> getTextures() {
+    ImmutableSet.Builder<ResourceLocation> builder = ImmutableSet.builder();
+
+    // Do the textures for interiors exist?
+
+    try {
+      ResourceLocation loc = interior;
+      loc = new ResourceLocation(loc.getResourceDomain(), "textures/" + loc.getResourcePath() + ".png");
+      Minecraft.getMinecraft().getResourceManager().getAllResources(loc);
+      loc = front;
+      loc = new ResourceLocation(loc.getResourceDomain(), "textures/" + loc.getResourcePath() + ".png");
+      Minecraft.getMinecraft().getResourceManager().getAllResources(loc);
+
+      // both found
+      builder.add(interior);
+      builder.add(front);
+      hasTextures = true;
+    } catch(IOException e) {
+      // either interior, front or both textures weren't found. we don't require them.
+      hasTextures = false;
+    }
+
+    return builder.build();
+  }
+
+  @Override
+  public IFlexibleBakedModel bake(IModelState state, VertexFormat format,
+                                  Function<ResourceLocation, TextureAtlasSprite> bakedTextureGetter) {
+    // bucket model
+    IFlexibleBakedModel bakedBucket = bucket.bake(bucket.getDefaultState(), format, bakedTextureGetter);
+    TRSRTransformation transform = state.apply(bucket);
+
+    return new BakedDynBucket(bakedBucket, transform, FluidRegistry.WATER, hasTextures);
```

Comment:
Might be good to implement IModelCustomData, to allow people to specify the fluid like that.